test/replication-luatest/gh_11180_wal_join_with_rollback_test.lua

```diff
@@ -0,0 +1,113 @@
+    cg.server:start()
+end)
+
+g.after_all(function(cg)
+    cg.server:drop()
+    if cg.replica then
+        cg.replica:drop()
+    end
+end)
+
+--
+-- gh-11180: some rollbacks could slip into a committed read-view due to the
+-- ignorance of rollbacks happening with volatile non-submitted journal entries,
+-- which were present in the read-view.
+--
+g.test_fetch_snapshot_during_rollback = function(cg)
+    --
+    -- Block WAL for the next entry, which will be a _gc_consumers txn for the
+    -- new replica.
+    --
+    cg.server:exec(function()
+        local s = box.schema.create_space('test')
+        s:create_index('pk')
+        box.error.injection.set('ERRINJ_WAL_DELAY_COUNTDOWN', 0)
+        t.assert_not(box.space._gc_consumers.index[0]:min())
+    end)
+    --
+    -- Start the new replica so it attempts to join and gets blocked.
+    --
+    cg.replica = server:new({
+        box_cfg = {
+            replication = cg.server.net_box_uri,
+            replication_timeout = 0.1,
+            replication_anon = true,
+            read_only = true,
+        }
+    })
+    cg.replica:start({wait_until_ready = false})
+    cg.server:exec(function()
+        local fiber = require('fiber')
+        local data = string.rep('a', 1000)
+        local timeout = 60
+        local s = box.space.test
+        --
+        -- The replica is trying to register the consumer already.
+        --
+        t.helpers.retrying({timeout = timeout}, function()
+            t.assert(box.space._gc_consumers.index[0]:min())
+            t.assert(box.error.injection.get('ERRINJ_WAL_DELAY'))
+        end)
+        --
+        -- Before a read-view is created for the replica, lets create some
+        -- dangling txns.
+        --
+        local function make_txn_fiber(id)
+            return fiber.create(function()
+                fiber.self():set_joinable(true)
+                s:replace{id, data}
+            end)
+        end
+        local function join_with_error(f, expected_err)
+            local ok, err = f:join()
+            t.assert_not(ok)
+            t.assert_covers(err:unpack(), expected_err)
+        end
+        local f1 = make_txn_fiber(1)
+        local f2 = make_txn_fiber(2)
+        local f3 = make_txn_fiber(3)
+        --
+        -- Let the _gc_consumers txn to commit, but block the new ones. The
+        -- read-view will then include them and will try to wait for them to get
+        -- committed.
+        --
+        box.error.injection.set('ERRINJ_WAL_DELAY_COUNTDOWN', 0)
+        box.error.injection.set('ERRINJ_WAL_DELAY', false)
+        t.helpers.retrying({timeout = timeout}, function()
+            t.assert(box.error.injection.get('ERRINJ_WAL_DELAY'))
+        end)
+        --
+        -- Txn1 went to WAL while txn2 and txn3 are waiting in the WAL queue.
+        -- Get them cancelled to trigger a cascading rollback which must be
+        -- spotted by the read-view creation code.
+        --
+        f2:cancel()
```

Comment:
AFAIU the issue with the current version of flush is that someone can jump in and cancel fiber with the request while we executing `fiber_sleep(0)`, otherwise the request should be sent to WAL and then synced by `wal_sync` (or request is cancelled even before we start flushing). Looks like the test does not reproduce this situation yet it failed with current version of flush. I did not find why though. Do you have ideas on this point?